library/bigip_software.py

```diff
@@ -212,673 +290,1060 @@
-    def __init__(self, *args, **kwargs):
-        self.result = dict(changed=False, changes=dict())
+        if link is None:
+            return None
 
-        self.current = dict()
+        if remote:
+            hotfix = self.handle_remote_source(link)
+            self._values['hotfix_name'] = hotfix
+            if link_md5:
+                hotfix_md5 = self.handle_remote_source(link_md5)
+                self._values['hfmd5_name'] = hotfix_md5
 
-        if kwargs['hotfix']:
-            kwargs['photfix'] = self.iso_info(kwargs['hotfix'])
+        else:
+            hotfix = os.path.split(link)[1]
+            self._values['hotfix_name'] = hotfix
 
-        if kwargs['software']:
-            kwargs['psoftware'] = self.iso_info(kwargs['software'])
+        return self._values['hotfix']
 
-        self.params = kwargs
+    @property
+    def options(self):
+        state = self._values['state']
+        create = self._values['create_volume']
+        opt = list()
 
-    def _find_iso_content(self, iso):
-        paths = ['/METADATA.XML', 'metadata.xml']
+        if create:
+            opt.append({'create-volume': True})
+        if state == 'activated':
+            opt.append({'reboot': True})
 
-        for path in paths:
-            try:
-                content = iso.record(path).content
-                return content
-            except KeyError:
-                pass
-        raise Exception(
-            "Unable to find metadata file in ISO. Please file a bug"
-        )
+        return opt
 
-    def iso_info(self, iso):
-        result = dict(
-            product=None,
-            version=None,
-            build=None
-        )
+    @property
+    def volume(self):
+        state = self._values['state']
+        volume = self._values['volume']
+        reuse = self._values['reuse_inactive_volume']
+
+        if state == 'activated' or state == 'installed':
+            if volume is not None:
+                if not self._volume_exists(volume):
+                    self._values['create_volume'] = True
+                    return self._values['volume']
+                else:
+                    self._check_active_volume(volume)
+                    return self._values['volume']
+
+            if volume is None:
+                if state == 'activated' or state == 'installed':
+                    if reuse:
+                        volume = self._get_next_available_volume()
+                        if self._volume_exists(volume):
+                            self._values['delete_volume'] = True
+                            self._values['create_volume'] = True
+                        else:
+                            self._values['create_volume'] = True
+                        self._values['volume'] = volume
+                        return self._values['volume']
+                    else:
+                        err = "You must specify a volume."
+                        raise F5ModuleError(err)
```

Comment:
I prefer error messages just part of the exception class. dont bother using a variable